src/syntaxtree.c

```diff
@@ -31,50 +31,101 @@
 #include <ctype.h>
 #include <stdarg.h>
 
+enum { MAX_ARITY = 4 };
+
 typedef Obj (*CompileFuncT)(Obj node, Expr expr);
 typedef Obj (*CompileArgT)(Expr expr);
+typedef Expr (*CodeFuncT)(Obj node);
+typedef Expr (*CodeArgT)(Obj node);
+
 
 typedef struct {
     const Char * argname;
     CompileArgT  argcomp;
+    CodeArgT     argcode;
+    Int          isStat;
 } ArgT;
 
 typedef struct {
     UInt1        tnum;
     CompileFuncT compile;
+    CodeFuncT    code;
     const Char * name;
-    UInt1        arity;
-    ArgT         args[4]; /* The maximum compiler arity is currently 4 */
+    UInt         arity;
+    ArgT args[MAX_ARITY]; /* The maximum compiler arity is currently 4 */
 } CompilerT;
 
 // We put compilers for statements and expressions into the same static array,
 // assuming that the set of their respective ids are disjoint.
 static const CompilerT Compilers[];
 #define COMPILER_ARITY(...) (sizeof((ArgT[]){ __VA_ARGS__ }) / sizeof(ArgT))
-#define COMPILER_GENERIC(tnum, compiler, ...)                                \
+#define COMPILER_GENERIC(tnum, compiler, coder, ...)                         \
     [tnum] = {                                                               \
-        tnum, compiler, #tnum, COMPILER_ARITY(__VA_ARGS__), { __VA_ARGS__ }  \
+        tnum,           compiler, coder, #tnum, COMPILER_ARITY(__VA_ARGS__), \
+        { __VA_ARGS__ }                                                      \
     }
 
-#define COMPILER(tnum, compiler) COMPILER_GENERIC(tnum, compiler)
+#define COMPILER(tnum, compiler, coder)                                      \
+    COMPILER_GENERIC(tnum, compiler, coder)
 
 #define COMPILER_(tnum, ...)                                                 \
-    COMPILER_GENERIC(tnum, SyntaxTreeDefaultCompiler, __VA_ARGS__)
+    COMPILER_GENERIC(tnum, SyntaxTreeDefaultCompiler,                        \
+                     SyntaxTreeDefaultCoder, __VA_ARGS__)
 
-#define ARG(name, func)                                                      \
+#define ARG(name, func, coder, isStat)                                       \
     {                                                                        \
-        name, func                                                           \
+        name, func, coder, isStat                                            \
     }
-#define ARG_(name) ARG(name, SyntaxTreeCompiler)
 
-#define ARGS(name) ARG(name, 0)
+#define ARG_EXPR(name, func, coder) ARG(name, func, coder, 0)
```

Comment:
Now would be a good time to also add some comments that explain what these macros do... What is `ARG_EXPR` about? Perhaps like this:
> // ARG_EXPR indicates an "expression argument", i.e., that the corresponding slot of the
> // statement or expression type it belongs to is meant to reference an expression.
> // This information is used for validation when coding an AST back into a T_BODY object.
> // The expression is stored in resp. read from the record entry <name>.

And then later:
> // ARG_STAT is like ARG_EXPR, except that it indicates that the corresponding slot should contain
> // a statement instead of an expression.

It's not great, but perhaps better than nothing. And it could be adapted to A